mobify/platform-scaffold
# Optimize Dependencies + Code Splitting

> _This PR will be updated with details of other optimizations as they are committed_

Try and get `vendor.js` as small as we possibly can!

**JIRA**: [WEB-1313](https://mobify.atlassian.net/browse/WEB-1313)
**Related PRs**: [progressive-web-sdk#566](https://github.com/mobify/progressive-web-sdk/pull/566), #476, #482

## Changes
- Alias some redundant dependencies
    - `babel-runtime`
    - many `lodash` methods
- Move small lodash modules to `dependencies` and add the full library under `devDependencies`

## How to test-drive this PR
- Run `MOBIFY_ANALYZE=true npm run prod:build`
- Verify the changes have had an impact
- Run `npm run dev`
- Navigate across the site and verify the site works as intended

## Todo
- [x] Code Splitting by route
- ~~Tree Shaking~~ (delayed to future)
- ~~Switch to Preact~~ (undesirable change)
- ~~Switch to React Fiber~~ (unsatisfactory savings at this time)
- ~~Switch to smaller alternatives of some of our current dependencies (Immutable.js, React-Form, others?) record thoughts on [those dependencies here](https://docs.google.com/document/d/1BdPX4QVVsuVjtUAr0Hj8z--zokkKbC1Lxm3PunI42dM/edit#heading=h.lzcyc5lyp9ce)~~ (effort and risks to make these changes is too large relative to the savings)
- ~~Explore [lodash-webpack-plugin](https://github.com/lodash/lodash-webpack-plugin)~~ (didn't get to this)
- [x] Remove `astro-client.js` from `main.js`

## Impact

> _The below impact notes are not all necessarily up to date with each other's changes. They are meant as a summary from that change's PR, to give a rough idea of what that change did. Expect variations in the details as all of these changes come together._

### Aliasing

`vendor.js` shrunk from `160.62kb` → **`154.01kb`**

`main.js` shrunk from `91.55kb` → `91.46` (negligible savings)

Total savings: **~`6.7kb`**

> _NOTE: the above savings were measured with [this update to the SDK](https://github.com/mobify/progressive-web-sdk/pull/566)_

| Before | After |
| --- | --- |
| ![vendor-before](https://cloud.githubusercontent.com/assets/734535/24274513/e078a906-0fe5-11e7-8386-433a3b5b1778.png) | ![vendor-after](https://cloud.githubusercontent.com/assets/734535/24274517/e8e2702c-0fe5-11e7-9f17-99c5aef2406d.png) |
| ![main-before](https://cloud.githubusercontent.com/assets/734535/24274521/ee6eec64-0fe5-11e7-8e05-3e28f9951c2d.png) | ![main-after](https://cloud.githubusercontent.com/assets/734535/24274525/f326bc1e-0fe5-11e7-944f-f65719b7bd9f.png) |


### Remove `astro-client.js` From `main.js`

Main.js shrunk from 70.93kb → **47.97kb**

Savings of **22.96kb**

| Before | After |
| --- | --- |
| ![before](https://cloud.githubusercontent.com/assets/734535/24476826/b1665e1c-1489-11e7-886b-5a91001c2689.png) | ![after](https://cloud.githubusercontent.com/assets/734535/24476828/b51f5d42-1489-11e7-80d0-e6e349f0d999.png) |


### Code Splitting

`main.js`: 48.93.kb → **31.35kb** gzip (~17kb smaller!)
`vendor.js`: 151.89kb → **146.12kb** gzip (~5kb smaller)

| Before | After |
| --- | --- |
| ![screen shot 2017-04-04 at 1 37 14 pm](https://cloud.githubusercontent.com/assets/4543395/24677495/db6f624c-193b-11e7-9362-58f6910a0fcb.png) | ![screen shot 2017-04-04 at 1 30 04 pm](https://cloud.githubusercontent.com/assets/4543395/24677310/33d434ea-193b-11e7-98b8-70cd4514418e.png) |
